File: base/action.py
import typing
from abc import ABC, abstractmethod

from PyQt5.QtXml import QDomDocument, QDomElement, QDomNode

import dfbase
import entities
import helper.forname
import syntax.parser
from dfexcept import DragonflyException
from output.console import Console


class Action(ABC):
	def __init__(self) -> None:
		self.__game = None
		self.__verb = None
		self.__sendingEvents = []

	@property
	def game(self) -> "dfbase.Game":
		return self.__game

	@game.setter
	def game(self, game: "dfbase.Game") -> None:
		self.__game = game

	@property
	def dictionary(self) -> "dfbase.Dictionary":
		return self.__game.dictionary

	@property
	def parser(self) -> "syntax.Parser":
		return self.__game.parser

	@property
	def verb(self) -> "entities.Verb":
		return self.__verb

	@verb.setter
	def verb(self, verb: "entities.Verb") -> None:
		self.__verb = verb

	@abstractmethod
	def init(self) -> bool:
		pass

	@abstractmethod
	def check(self) -> bool:
		pass

	@abstractmethod
	def carryOut(self) -> None:
		pass

	@abstractmethod
	def report(self) -> None:
		pass

	def execute(self) -> None:
		self.__sendingEvents.clear()

		# Initialize action
		if not self.init(): return None

		# BEFORE EVENT
		for n in self.__sendingEvents:
			if not n.doBefore(self): return None

		# Check action
		if not self.check(): return None

		self.__sendingEvents.clear()

		# Action performs the changes
		self.carryOut()

		# AFTER EVENT
		for n in self.__sendingEvents:
			if not n.doAfter(self): return None

		# Report the result
		self.report()

	def sendEventLater(self, noun: "entities.Noun") -> None:
		self.__sendingEvents.append(noun)

	def fireResponse(self, id: str) -> bool:
		if not self.__verb.hasResponse(id):
			raise DragonflyException(f"Verb {self.__verb} has not response: {id}.")

		Console.println(self.__verb.getResponse(id))
		return False

class ActionEvent:
	def __init__(self) -> None:
		self.__actions = []
		self.__cancel = False

		self.__responses = []
		self.__conditions = []

	def __str__(self) -> str:
		lst = []
		for a in self.__actions:
			lst.append(str(a))

		return ", ".join(lst)

	@property
	def actions(self) -> typing.List:
		return self.__actions

	@property
	def cancel(self) -> bool:
		return self.__cancel

	def match(self, action: Action) -> bool:
		return action.__class__ in self.__actions

	def addResponse(self, response: "ActionResponse") -> None:
		self.__responses.append(response)

	def addCondition(self, condition: "Condition") -> None:
		self.__conditions.append(condition)

	def checkConditions(self, action: Action) -> bool:
		check = True
		for cond in self.__conditions:
			if not cond.check(action): check = False

		return check

	def execute(self, action: Action) -> None:
		for r in self.__responses:
			r.execute(action)


	def load(self, element: QDomElement) -> None:
		# Load action list
		for a in element.attribute("actions").split(","):
			aClass = helper.forname.getClass(a.strip(), defaultModule="actions")
			self.__actions.append(aClass)
			
		self.__cancel = element.attribute("cancel") == "true"
		
		# Load responses and conditions
		for i in range(element.childNodes().count()):
			child = element.childNodes().at(i)
			if child.nodeType() == QDomNode.ElementNode:
				e = child.toElement()
				if e.nodeName() == "response":
					responseClass = helper.forname.getClass(e.attribute("class"), defaultModule="responses")
					response = responseClass()
					response.load(e)
					self.addResponse(response)

				if e.nodeName() == "if":
					condClass = helper.forname.getClass(e.attribute("class"), defaultModule="conditions")
					cond = condClass()
					cond.load(e)
					self.addCondition(cond)

	def save(self, doc: QDomDocument, nodeName: str) -> QDomElement:
		element = doc.createElement(nodeName)
		
		# Save action class names
		aNames = []
		for a in self.__actions:
			aNames.append(a.__name__)

		element.setAttribute("actions", ", ".join(aNames))

		# Cancel
		element.setAttribute("cancel", "true" if self.__cancel else "false")

		# Conditions
		for c in self.__conditions:
			element.appendChild(c.save(doc))

		# Responses
		for r in self.__responses:
			element.appendChild(r.save(doc))

		return element

		
class ActionResponse(ABC):
	def __init__(self) -> None:
		pass

	@abstractmethod
	def execute(self, action: Action) -> None:
		pass

	@abstractmethod
	def load(self, element: QDomElement) -> None:
		pass

	@abstractmethod
	def save(self, doc: QDomDocument) -> QDomElement:
		element = doc.createElement("response")
		element.setAttribute("class", self.__class__.__name__)
		return element

class Condition(ABC):
	def __init__(self) -> None:
		pass

	@abstractmethod
	def check(self, action: Action) -> bool:
		pass

	@abstractmethod
	def load(self, element: QDomElement):
		pass

	@abstractmethod
	def save(self, doc: QDomDocument) -> QDomElement:
		element = doc.createElement("if")
		element.setAttribute("class", self.__class__.__name__)

		return element

File: base/action.js
/**
 * Represents an action performed by the player.
 * When the verb is parsed, the associated action is executed by the execute() method.
 */


import { Output } from "./output.js"
import { Noun } from "./noun.js";

/**
 * Represents an action performed by the player.
 * When the verb is parser, the associated action is executed by execute() method.
 *
 * @export
 * @class Action
 */
export class Action {
	/**
	 * Creates an instance of Action.
	 * @memberof Action
	 */
	constructor() {
		this.book = null;
		this.verb = null;
		this.sendingEvents = [];
	}

	/**
	 * Initialize the action. Behavior depends on the type of action.
	 * Direct and indirect objects are usually looked up in the dictionary and 
	 * if these objects are not found, False is returned for cancel the action.
	 *
	 * @return {boolean} False: causes the action to be cancelled.
	 * @abstract
	 * @memberof Action
	 */
	async init() {
		return false;
	}

	/**
	 * The action check consists in verifying if an object
	 *	complies with the norm of the action. Example:
	 *	LookInside checks if the direct object is not "closed".
	 *	"the player will not be able to see inside the chest if it is closed"
	 *
	 * @return {boolean} causes the action to be cancelled.
	 * @abstract
	 * @memberof Action
	 */
	async check() {
		return false;
	}

	/**
	 * Perform the action. Example:
	 * If the action is TakeObject, the direct object is moved to player's inventory.
	 * @abstract
	 * @memberof Action
	 */
	async carryOut() {

	}

	/**
	 * Report occurs after when action has been performed. Usually show a message.
	 * Example:
	 * "you get a flashlight"
	 *
	 * @abstract
	 * @memberof Action
	 */
	async report() {

	}

	/**
	 * The execute() method has the following steps:
	 * 1)	Abstract method init(): initialize the action. If False is
	 * 	returned from init(), action is cancelled.
	 * 1a)	Execute before event on objects handled by the action. If False is returned
	 * 	from doBefore() method, action is cancelled.
	 * 2)	Abstract method check(): check the action. False: action is cancelled.
	 * 3)	Abstract method carryOut(): perform the action.
	 * 3b)	Execute after event on objects handled by the action. If False is returned
	 * 	from doAfter() method, action is cancelled in this point.
	 * 4)	Abstract method report(): report the action after performed.
	 *
	 * @memberof Action
	 */
	async execute() {
		this.sendingEvents = []

		// Initialize action
		if (!await this.init()) return ;

		// BEFORE EVENT
		for (const n of this.sendingEvents) {
			if (!await n.doBefore(this)) return ;
		}

		// Check action
		if (!await this.check()) return ;

		this.sendingEvents = [];

		// Action performs the changes
		await this.carryOut();

		// AFTER EVENT
		for (const n of this.sendingEvents) {
			if (!await n.doAfter(this)) return ;
		}

		// Report the result
		await this.report();
	}

	/**
	 * Add the target object to list. This list will be used to send before
	 * and after events.
	 *
	 * @param {Noun} noun Target object
	 * @memberof Action
	 */
	sendEventLater(noun) {
		this.sendingEvents.push(noun);
	}

	/**
	 * Execute a verb's response. The response may exists in the verb and show
	 * the response message. The verb is associated with the current action.
	 *
	 * @param {string} id response id declared in the verb.
	 * @return {boolean} ALWAYS returns false for programming convenience.
	 * @memberof Action
	 */
	fireResponse(id) {
		if (!this.verb.hasResponse(id)) {
			Output.error(`the verb "${this.verb.getName()}" does not have response "${id}"`);
			return false;
		}

		Output.print(this.verb.getResponse(id));

		return false;
	}
	

	/**
	 * Return a list of all responses that may occurs.
	 *
	 * @return {Array.<string>}
	 * @abstract
	 * @memberof Action
	 */
	responses() {
		return [];
	}

	/**
	 * Return action string info.
	 *
	 * @return {string} string of action.
	 * @memberof Action
	 */
	toString() {
		return this.constructor.name;
	}
};

/**
 * Single action takes no arguments. Usually represents verb like "jump", "scream".
 *	By default, SingleAction fires response: "nothing-happens"
 *
 *	Responses:
 *	nothing-happens

 * @export
 * @class SingleAction
 * @extends {Action}
 */
 export class SingleAction extends Action {

	constructor() {
		super();
	}

	init() {
		return true;
	}

	check() {
		return true;
	}

	carryOut() {
		
	}

	report() {
		this.fireResponse("nothing-happens");
	}

	responses() {
		return ["nothing-happens"];
	}
}

/**
 * DefaultAction takes one argument (direct object).
 * By default, fires response "nothing-happens"
 * Responses:
 * direct-not-found
 * direct-is-the-player
 * nothing-happens
 *
 * @export
 * @class DefaultAction
 * @extends {Action}
 */
export class DefaultAction extends Action {

	constructor() {
		super();
	}

	async init() {
		let lst = this.book.player.children(this.book.parser.directObjectString);
		lst = lst.concat(this.book.player.container.children(this.book.parser.directObjectString));

		if (lst.length === 0) {
			return this.fireResponse("direct-not-found");
		}

		this.book.parser.directObject = await this.book.dictionary.objectChooserDialog.execute(lst);
		if (!this.book.parser.directObject) {
			return false;
		}

		this.sendEventLater(this.book.parser.directObject);

		return true;
	}

	async check() {
		if (this.book.parser.directObject == this.book.player) {
			return this.fireResponse("direct-is-the-player")
		}
		return true;
	}

	async carryOut() {
		
	}

	async report() {
		this.fireResponse("nothing-happens")
	}

	responses() {
		return ["nothing-happens", "direct-not-found", "direct-is-the-player"]
	}
}
